torch/csrc/distributed/rpc/tensorpipe_agent.h

```diff
@@ -0,0 +1,111 @@
+
+#include <atomic>
+#include <thread>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+struct TensorPipeRpcBackendOptions : public RpcBackendOptions {
+  std::map<std::string, worker_id_t> workerNameToId;
+};
+
+class TensorPipeAgent : public RpcAgent {
+ public:
+  TensorPipeAgent(
+      worker_id_t selfId,
+      std::string selfName,
+      std::shared_ptr<::c10d::Store> addressStore,
+      TensorPipeRpcBackendOptions opts);
+
+  TensorPipeAgent(const TensorPipeAgent&) = delete;
+  TensorPipeAgent& operator=(const TensorPipeAgent&) = delete;
+
+  std::shared_ptr<FutureMessage> send(const WorkerInfo& to, Message&& message)
+      override;
+
+  // join() and sync() would be deprecated -
+  // https://github.com/pytorch/pytorch/issues/27647
+  void join() override;
+  void sync() override;
+  void startImpl() override;
+  void shutdownImpl() override;
+
+  ~TensorPipeAgent() override;
+
+  const WorkerInfo& getWorkerInfo(const std::string& workerName) const override;
+  const WorkerInfo& getWorkerInfo(worker_id_t workerId) const override;
+  std::vector<WorkerInfo> getWorkerInfos() const override;
+
+  std::unordered_map<std::string, std::string> getMetrics() override {
+    std::unordered_map<std::string, std::string> metrics;
+    return metrics;
+  }
+
+  void addGilWaitTime(const std::chrono::microseconds gilWaitTime) override {}
+
+ private:
+  const std::string& findWorkerURL(const WorkerInfo& worker) const;
+
+  std::string createUniqueShmAddr();
+
+  // Read function used both to read responses to requests and to read incoming
+  // messages.
+  void pipeRead(
+      std::shared_ptr<tensorpipe::Pipe>,
+      std::function<void(const tensorpipe::Error&, Message&&)>);
+
+  // Write message to pipe
+  void pipeWrite(
+      std::shared_ptr<tensorpipe::Pipe>,
+      Message&& message,
+      std::function<void(const tensorpipe::Error&)>);
+
+  // Callback of listener accept()
+  void onListenerAccepted(
+      const tensorpipe::Error& error,
+      std::shared_ptr<tensorpipe::Pipe>& pipe);
+
+  // Respond to a call from a peer
+  void respond(std::shared_ptr<tensorpipe::Pipe>& pipe);
+
+  void sendCompletedResponseMessage(
+      std::shared_ptr<tensorpipe::Pipe>& pipe,
+      std::shared_ptr<FutureMessage>& futureResponseMessage,
+      const c10::optional<utils::FutureError>& error,
+      uint64_t messageId);
+
+  std::shared_ptr<tensorpipe::Context> context_;
+  std::shared_ptr<tensorpipe::Listener> listener_;
+  std::unordered_map<worker_id_t, std::shared_ptr<tensorpipe::Pipe>>
+      connectedPipes_;
+
+  // We need map one keyed on name and one on id for easy lookup.
+  std::unordered_map<worker_id_t, WorkerInfo> workerIdToInfo_;
+  std::unordered_map<std::string, WorkerInfo> workerNameToInfo_;
+  std::unordered_map<std::string, std::string> workerNameToURL_;
+
+  const std::shared_ptr<::c10d::Store> addressStore_;
+  const TensorPipeRpcBackendOptions opts_;
+  std::atomic<bool> isRunning_;
```

Comment:
Can we reuse `rpcAgentRunning_` from RpcAgent instead of having another flag here? cc @osalpekar

Reply:
Yes, we can just use `rpcAgentRunning_` wherever needed. `start` and `shutdown` in `RpcAgent` take care of setting and unsetting this flag.